flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,333 @@
+
+	private static File haStorageDir;
+
+	private static TestingServer zkServer;
+
+	private static MiniClusterResource miniClusterResource;
+
+	private static OneShotLatch waitForCheckpointLatch = new OneShotLatch();
+	private static OneShotLatch failInCheckpointLatch = new OneShotLatch();
+	private static OneShotLatch successfulRestoreLatch = new OneShotLatch();
+
+	@BeforeClass
+	public static void setup() throws Exception {
+		zkServer = new TestingServer();
+
+		Configuration config = new Configuration();
+		config.setInteger(ConfigConstants.LOCAL_NUMBER_JOB_MANAGER, NUM_JMS);
+		config.setInteger(ConfigConstants.LOCAL_NUMBER_TASK_MANAGER, NUM_TMS);
+		config.setInteger(TaskManagerOptions.NUM_TASK_SLOTS, NUM_SLOTS_PER_TM);
+
+		haStorageDir = temporaryFolder.newFolder();
+
+		config.setString(HighAvailabilityOptions.HA_STORAGE_PATH, haStorageDir.toString());
+		config.setString(HighAvailabilityOptions.HA_CLUSTER_ID, UUID.randomUUID().toString());
+		config.setString(HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, zkServer.getConnectString());
+		config.setString(HighAvailabilityOptions.HA_MODE, "zookeeper");
+
+		// we have to manage this manually because we have to create the ZooKeeper server
+		// ahead of this
+		miniClusterResource = new MiniClusterResource(
+			new MiniClusterResource.MiniClusterResourceConfiguration(
+				config,
+				NUM_TMS,
+				NUM_SLOTS_PER_TM));
+
+		miniClusterResource.before();
+	}
+
+	@AfterClass
+	public static void tearDown() throws Exception {
+		miniClusterResource.after();
+
+		zkServer.stop();
+		zkServer.close();
+	}
+
+	/**
+	 * Verify that we don't start a job from scratch if we cannot restore any of the
+	 * CompletedCheckpoints.
+	 *
+	 * <p>Synchronization for the different steps and things we want to observe happens via
+	 * latches in the test method and the methods of {@link CheckpointBlockingFunction}.
+	 *
+	 * <p>The test follows these steps:
+	 * <ol>
+	 *     <li>Start job and block on a latch until we have done some checkpoints
+	 *     <li>Block in the special function
+	 *     <li>Move away the contents of the ZooKeeper HA directory and make it non-writable
+	 *       to make creating and restoring from checkpoints impossible
+	 *     <li>Unblock the special function, which now induces a failure
+	 *     <li>Make sure that the job does not recover successfully
+	 *     <li>Move back the HA directory
+	 *     <li>Make sure that the job recovers, we use a latch to ensure that the operator
+	 *       restored successfully
+	 * </ol>
+	 */
+	@Test(timeout = 120_000L)
+	public void testRestoreBehaviourWithFaultyStateHandles() throws Exception {
+		CheckpointBlockingFunction.allowedInitializeCallsWithoutRestore.set(1);
+		CheckpointBlockingFunction.successfulRestores.set(0);
+		CheckpointBlockingFunction.illegalRestores.set(0);
+		CheckpointBlockingFunction.afterMessWithZooKeeper.set(false);
+		CheckpointBlockingFunction.failedAlready.set(false);
+
+		waitForCheckpointLatch = new OneShotLatch();
```

Comment:
given that there is only a single test we could also make the static latches final.

Reply:
I did it like this because it allows running the test in a loop in IntelliJ, which doesn't start a new JVM for each execution, meaning the static fields are not re-initialized.

Reply:
i see, that makes sense.